contests/_24-PDP/c-souvlakia-solution.md

```diff
@@ -0,0 +1,223 @@
+Πολυπλοκότητα Dijkstra $$\mathcal{O}((N+M)\cdot log{N})$$ 
+και απαντήσεων ερωτημάτων $$\mathcal{O}(N^2)$$. 
+Η συνολική πολυπλοκότητα περιορίζεται από τον χειρότερο όρο άρα είναι της τάξης $$\mathcal{O}(N^2)$$ [^1].
+Παρακάτω δίνεται μία ενδεικτική υλοποίηση αυτής της λύσης.
+
+{% include code.md solution_name='souvlakia_n2.cc' %}
+
+## Mέτρια λύση - $$\mathcal{O}(N \cdot log(D \cdot N))$$
+
+Γνώσεις που θα χρειαστούμε: ([Segment Tree](https://kallinikos.github.io/Segment-Trees)) ή ([Binary Indexed Tree](https://kallinikos.github.io/Binary-Indexed-Tree))
+
+Η δυσκολία που παρουσιάζει το πρόβλημα είναι ότι έχουμε τρείς διαφορετικές συνθήκες ανεξάρτητες μεταξύ τους. Θα πρέπει 
+να μπορέσουμε να αφαιρέσουμε τουλάχιστο τη μία από τις τρεις συνθήκες με κάποιο έξυπνο τρόπο ώστε να μπορέσουμε να δουλέψουμε
+τις άλλες δυο. 
+Εφόσον έχουμε να βρούμε διάταξη μεταξύ των αντίστοιχων τιμών, μπορούμε να αφαιρέσουμε την μια από τις τρεις συνθήκες αντικαθιστώντας την με 
+τον "χρόνο". Δηλαδή αν ταξινομήσουμε όλα τα καταστήματα με τη μία παράμετρο, έστω τη $$x$$ και κάνουμε προσπέλαση από το μικρότερο προς το 
+μεγαλύτερο, τότε τη χρονική στιγμή που θα ελέγχουμε το κατάστημα $$i$$ έχουμε προσπελάσει όλα τα καταστήματα $$j$$ 
+με $$x_j\lt x_i$$. Αυτό που καταφέρνουμε είναι ότι όλα τα καταστήματα που μπορούν να βγάλουν το κατάστημα $$i$$ ως 
+μη επιλέξιμο, θα τα έχουμε 
+προσπεράσει με τη σάρωση μας πριν το $$i$$. Όσα καταστήματα βρίσκονται μετά από το $$i$$ δεν το επηρεάζουν.
+
+Αυτό βέβαια αυτομάτως σημαίνει ότι δεν μπορούμε να απαντήσουμε τα queries με τη σειρά που μας δίνονται 
+και θα πρέπει να τα προϋπολογίσουμε με τη σειρά που μας βολεύει και να τα απαντήσουμε offline.  
+Έχοντας ταξινομήσει ως προς $$x$$ ξεχνάμε τελείως τις τιμές αυτές του $$x$$ και επικεντρονόμαστε στις άλλες 
+δυο τιμές τις $$y$$ και $$z$$.
+Για κάθε $$y_i$$ που βρίσκουμε καθώς ελέγχουμε το κατάστημα $$i$$ θέλουμε να ξέρουμε αν υπάρχει ένα τουλάχιστο κατάστημα $$j$$ με $$y_j \lt y_i$$ 
+και $$z_j \lt z_i$$. 
+
+Ας το εκφράσουμε λίγο διαφορετικά: αν διατηρούμε από τα καταστήματα που ελέγξαμε έως τώρα με $$y_j \lt y_i$$ μόνο το 
+$$z_{min}$$, 
+και αυτό το $$z_{min} \lt z_i$$ τότε το κατάστημα 
+$$i$$ δεν είναι επιλέξιμο γιατί τουλάχιστο ένα άλλο κατάστημα προτιμάται. Η δομή αυτή που ζητάμε θα πρέπει να μας απαντά 
+το ελάχιστο $$z$$ για όλα $$y \in [0,y_i)$$ 
+δηλαδή να μας δίνει το __Range Minimum Query__ όλων των τιμών $$z$$ σε κάποιο range $$[0,y_i-1]$$. 
+Μια εύκολη λύση είναι να χρησιμοποιήσουμε ένα Segment Tree ή ένα Binary Indexed Tree. 
+
+Δυστυχώς οι λύσεις αυτές ενώ χρονικά είναι ικανοποιητικές, δεν θα περνούσαν τα αυστηρά όρια μνήμης του 
+προβλήματος ακόμα και αν επαναχρησιμοποιήσουμε τη μνήμη που δεν χρειάζεται[^2]. 
+Το οικονομικότερο σε μνήμη 
+και ταχύτερο Binary Indexed Tree θα ήθελε μνήμη της τάξης $$\mathcal{O}(D\cdot N)$$ (όπου $$D$$ το μέγιστο βάρος μιας edge του γράφου), ενώ 
+το segment tree θα ήθελε μνήμη τάξης $$\mathcal{O}(2^{ceil(log_2(D\cdot N))+1})$$.
```

Comment:
O(DN)

Reply:
Δεν κατάλαβα. Χωρίς την τελεία του πολλαπλασιασμού; Δεν θα μπερδεύει;

Reply:
Εννοώ ότι και το segment tree έχει μνήμη O(DN), απλά στην περίπτωση του BIT είναι DN, ενώ στο segment tree είναι περίπου 4DN.

Reply:
Η τελεία είναι χρήσιμη

Reply:
To 4DN είναι μια απαισιοδοξη προσέγγυση. Το μεγεθος είναι το διπλάσιο της στρογγυλοποιησης στην επόμενη δυναμη του 2. Θα το βαλω σαν υποσημειωση τον πραγματισμο υπολογισμο μνημης.

Reply:
Για DN = 2^n + 1, έχετε ότι 2^(ceil(log2(2^n + 1)) + 1) = 2^(ceil(n + ε) + 1) = 2^(n + 1 + 1) = 4 * 2^n  = 4 * DN -  4. 
που δεν απέχει πολύ από το DN. Άμα τα testcases ήταν ισχυρά, δεν θα είχε διαφορά ποια φόρμουλα χρησιμοποιούμε. Εδώ παρατηρείτε κάποια διαφορά; 

(Για DN = 2^n, από την άλλη κερδίζουμε 2x)

Reply:
Με 64Mb θα είμασταν χαμένοι απο χερι με "γεματα" test cases (αν έφταναν στο maxD*maxN). 
Το implicit με static allocation πλησιάζει τα 64Mb.

Reply:
O(2^{ceil(log_2(D\cdot N))+1}) = O(DN), άρα η πρόταση δεν δείχνει τη διαφορά της μνήμης μεταξύ του BIT και του segment tree.

Για αρκετές τιμές των DN, η φόρμουλα 2^{ceil(log_2(D\cdot N))+1} είναι κοντά στο 4DN ακόμα και όταν δεν είμαστε στα γεμάτα test cases. Σκέφτομαι μήπως αυτή η φόρμουλα περιπλέκει τα πράγματα χωρίς λόγο.

Reply:
Την σε υποσημειωση. Στον κωδικα μπορει να αλλαχθει, αλλα δεν περιμενω να μαθει καποιος segment tree απο μια εφαρμογη του. Θα το διαβασει απο αλλου...μαζι με τη θεωρεια